smol/capp/generate/groundstate/solver.py

```diff
@@ -0,0 +1,553 @@
+    corresponds only to an upper bound of the global ground state of the corresponding
+    infinitely sized system.
+
+    For more details on the nature of the global problem, finding upper and lower bounds
+    please see:
+
+    https://doi.org/10.1103/PhysRevB.94.134424
+    """
+
+    def __init__(
+        self,
+        ensemble: Ensemble,
+        initial_occupancy: ArrayLike = None,
+        fixed_composition: ArrayLike = None,
+        other_constraints: list = None,
+        term_coefficients_cutoff: float = 0.0,
+        warm_start: bool = False,
+        solver: str = "SCIP",
+        solver_options: dict = None,
+    ):
+        """Initialize GroundStateSolver.
+
+        Args:
+            ensemble (Ensemble):
+                An ensemble to initialize the problem with. If you want to modify
+                sub-lattices, please do that in ensemble before initializing the
+                groundstate. Sub-lattices can not be modified after giving the ensemble!
+            initial_occupancy (ArrayLike): optional
+                An initial occupancy used to set the occupancy of manually restricted
+                sites that may have more than one allowed species. Also used to set up
+                the initial composition when solving in a canonical ensemble.
+                Must be provided if any site has been manually restricted, or solving
+                canonical ensemble.
+            fixed_composition (ArrayLike): optional
+                An array of floats corresponding to the "counts" format of composition
+                in CompositionSpace. See moca.composition. Unit is per super-cell.
+                If the ensemble is canonical (no chemical potential has been set), will
+                need to constrain the ground-state problem to a fixed composition.
+                If initial_occupancy is provided while fixed_composition is None,
+                composition will be fixed to that of initial_occupancy.
+                Will be ignored in a semi-grand ensemble.
+                If any of initial_occupancy or fixed_composition is provided, you are
+                fully responsible to ensure that they satisfy charge balance and other
+                constraints.
+            other_constraints (list): optional
+                Other constraints to set in composition space. See moca.composition
+                for detailed description.
+                Note that constraints are now to be satisfied with the number of sites
+                in sub-lattices of the supercell instead of the primitive cell.
+            term_coefficients_cutoff (float): optional
+                Minimum cutoff to the coefficient of terms. If the absolute value of a
+                term coefficient is less than cutoff, it will not be included in the
+                objective function. If no cutoff is given, will include every term in the
+                energy objective function.
```

Comment:
Is this cutoff ever used? And what exactly is it for? is it to ignore some entries in the interaction tensors that are below the cutoff?

Reply:
This will be used. After a polynomial boolean function has been generated as the objective, we want to drop the function terms that has less than cutoff absolute value to reduce the problem size as much as possible. This cutoff is not set to specifically ignore some interaction tensor, because this cutoff leverages the truncation of terms in the FINAL SIMPLIFIED objective function, which also includes other interactions such as Ewald.

Reply:
I should make the DOC strings more evident though.

Reply:
Changed doc string.